projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/experiment/Experiment.java

```diff
@@ -0,0 +1,340 @@
+   * <p>
+   * <b>Note</b>: If a variable with the provided name already exists, the creation will fail with
+   * an {@link ExperimentalVariableExistsException} and no variable is added to the design. You can
+   * check via {@link Result#isFailure()} if this is the case.
+   *
+   * @param variableName a declarative and unique name for the variable
+   * @param levels       a list containing at least one value for the variable
+   * @return a {@link Result} object containing the {@link VariableName} or contains declarative
+   * exceptions. The result will contain an {@link ExperimentalVariableExistsException} if the
+   * variable already exists or an {@link IllegalArgumentException} if no level has been provided.
+   * @since 1.0.0
+   */
+  public Result<VariableName, Exception> addVariableToDesign(String variableName,
+      List<ExperimentalValue> levels) {
+    if (levels.size() < 1) {
+      return Result.failure(new IllegalArgumentException(
+          "At least one level required. Got " + levels));
+    }
+
+    if (experimentalDesign.isVariableDefined(variableName)) {
+      return Result.failure(new ExperimentalVariableExistsException(
+          "A variable with the name " + variableName + " already exists."));
+    }
+    try {
+      ExperimentalVariable variable = ExperimentalVariable.createForExperiment(this, variableName,
+          levels.toArray(ExperimentalValue[]::new));
+      experimentalDesign.variables.add(variable);
+      return Result.success(variable.name());
+    } catch (IllegalArgumentException e) {
+      return Result.failure(e);
+    }
+  }
+
+  /**
+   * Adds a variable to the design with the levels specified. If the variable exists already then
+   * adds the levels to the variable.
+   *
+   * @param variableName the name of the variable
+   * @param levels       the levels of the variable
+   */
+  private void addVariableOrLevels(String variableName, List<ExperimentalValue> levels) {
+    addVariableToDesign(variableName, levels).ifFailure(
+        e -> {
+          if (e instanceof ExperimentalVariableExistsException) {
+            // at this point we know there is a variable with the name `specimen`, so we only need to add the levels
+            for (ExperimentalValue level : levels) {
+              experimentalDesign.addLevelToVariable(variableName, level).ifFailure(
+                  e2 -> {
+                    //FIXME what exception to throw here?
+                    //   we know that the experimental variable must be defined at that point so we do not expect this exception
+                    throw new RuntimeException(
+                        "could not add level " + level + " to variable "
+                            + variableName, e2);
+                  });
+            }
+          } else if (e instanceof IllegalArgumentException) {
+            //FIXME what exception to throw here?
+            throw new RuntimeException(e);
+          }
```

Comment:
I find these indentations are hard to read. I also think that the function should do only one thing, the name already says otherwise